File: webflux-dialogue/src/main/java/com/study/webflux/rag/application/dialogue/service/DialogueMessageService.java
package com.study.webflux.rag.application.dialogue.service;

import java.util.ArrayList;
import java.util.List;

import org.springframework.stereotype.Service;

import com.study.webflux.rag.domain.dialogue.model.ConversationContext;
import com.study.webflux.rag.domain.dialogue.model.Message;
import com.study.webflux.rag.domain.memory.model.MemoryRetrievalResult;
import com.study.webflux.rag.domain.retrieval.model.RetrievalContext;

@Service
public class DialogueMessageService {

	private final SystemPromptService systemPromptService;

	public DialogueMessageService(SystemPromptService systemPromptService) {
		this.systemPromptService = systemPromptService;
	}

	public List<Message> buildMessages(RetrievalContext context,
		MemoryRetrievalResult memories,
		ConversationContext conversationContext,
		String currentQuery) {
		List<Message> messages = new ArrayList<>();

		String fullSystemPrompt = systemPromptService.buildSystemPrompt(context, memories);
		if (fullSystemPrompt == null || fullSystemPrompt.isBlank()) {
			fullSystemPrompt = "You are a helpful assistant.";
		}
		messages.add(Message.system(fullSystemPrompt));

		conversationContext.turns().stream().filter(turn -> turn.response() != null)
			.forEach(turn -> {
				messages.add(Message.user(turn.query()));
				messages.add(Message.assistant(turn.response()));
			});

		messages.add(Message.user(currentQuery));

		return messages;
	}
}

File: app-miyou/src/main/kotlin/com/miyou/app/application/dialogue/pipeline/stage/DialogueMessageService.kt
package com.miyou.app.application.dialogue.pipeline.stage

import com.miyou.app.domain.dialogue.model.ConversationContext
import com.miyou.app.domain.dialogue.model.Message
import com.miyou.app.domain.dialogue.model.PersonaId
import com.miyou.app.domain.memory.model.MemoryRetrievalResult
import com.miyou.app.domain.retrieval.model.RetrievalContext
import org.springframework.stereotype.Service

@Service
class DialogueMessageService(
    private val systemPromptService: SystemPromptService,
) {
    companion object {
        private const val DEFAULT_SYSTEM_PROMPT = "You are a helpful assistant."
    }

    fun buildMessages(
        personaId: PersonaId,
        context: RetrievalContext,
        memories: MemoryRetrievalResult,
        conversationContext: ConversationContext,
        currentQuery: String,
    ): List<Message> {
        require(currentQuery.isNotBlank()) { "Current query must not be null or blank." }

        val messages =
            buildList {
                val fullSystemPrompt = systemPromptService.buildSystemPrompt(personaId, context, memories)
                add(Message.system(fullSystemPrompt.ifBlank { DEFAULT_SYSTEM_PROMPT }))

                conversationContext.turns.forEach { turn ->
                    if (turn.response != null) {
                        add(Message.user(turn.query))
                        add(Message.assistant(turn.response))
                    }
                }

                add(Message.user(currentQuery))
            }
        return messages
    }
}
